vmware_cluster_ha: Add APD and PDL configuration

mariolenz opened the pull request:
Depends-On: #495

##### SUMMARY
Adding "all paths down" and "permanent device loss" config options.
 
Fixes #451

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
vmware_cluster_ha

mariolenz commented:
recheck

mariolenz commented:
recheck

mariolenz commented:
@goneri 
Is there something wrong with the CI pipeline? I always get `ERROR: Could not open requirements file: [Errno 2] No such file or directory: '/home/zuul/.ansible/collections/ansible_collections/community/vmware/requirements.txt'` in [ansible-test-cloud-integration-govcsim-python36_1_of_3](https://dashboard.zuul.ansible.com/t/ansible/build/1b1a4550edc24510b2ded9bc8ff3ed28).

I think your last PR has the same problem.

Akasurde commented on tests/integration/targets/vmware_cluster_ha/tasks/main.yml:
```diff
@@ -195,6 +195,70 @@
+      datacenter_name: "{{ dc1 }}"
+      cluster_name: test_cluster_ha
+      apd_response: 'restartAggressive'
+    check_mode: yes
```
```suggestion
    check_mode: True
```

Akasurde commented on tests/integration/targets/vmware_cluster_ha/tasks/main.yml:
```diff
@@ -195,6 +195,70 @@
+      that:
+        - change_pdl_response.changed
+
+  - name: Change PDL response to "restartAggressive" again
```
👍

Akasurde commented:
> @goneri
> Is there something wrong with the CI pipeline? I always get `ERROR: Could not open requirements file: [Errno 2] No such file or directory: '/home/zuul/.ansible/collections/ansible_collections/community/vmware/requirements.txt'` in [ansible-test-cloud-integration-govcsim-python36_1_of_3](https://dashboard.zuul.ansible.com/t/ansible/build/1b1a4550edc24510b2ded9bc8ff3ed28).
> 
> I think your last PR has the same problem.

This caused by https://github.com/ansible/ansible-zuul-jobs/pull/697/files#diff-14585baa1321c2341f4ebad0340f6631ed654741aa3491ebcd50e34eb4a65031R50

I raised a PR https://github.com/ansible/ansible-zuul-jobs/pull/699 as workaround. You can add 

```
Depends-On: <PR>
```
in PR description and `re-check` to re-trigger the CI run.

mariolenz commented:
re-check

mariolenz commented:
recheck

mariolenz commented:
@goneri
> You can add
> 
> ```
> Depends-On: <PR>
> ```
> 
> in PR description and `re-check` to re-trigger the CI run.

I still get the same error. Did I do something wrong or is this because ansible/ansible-zuul-jobs#699 fails some tests and is therefore ignored?

pabelanger commented:
recheck

pabelanger commented:
recheck

mariolenz commented:
recheck

mariolenz commented:
recheck

mariolenz commented:
recheck

mariolenz commented:
recheck

mariolenz commented:
recheck

mariolenz commented:
[for the record](https://dashboard.zuul.ansible.com/t/ansible/buildset/e3297a26dad5408a973b6cefd62bbc6f)

mariolenz commented:
recheck

mariolenz commented:
@Akasurde @sky-joker @Tomorrow9 

Sorry, this PR has been quite a mess. To begin with, there were a couple of problems with the CI pipeline that took some time to fix. Then I accidentally pushed broken integration tests. I fixed this, but after PR #448 has been merged there was a merge conflict. Well, I fixed that but in the meantime PR #478 was merged which resulted in _another_ merge conflict. Fixing this resulted in a CI error I didn't understand, but after a `recheck` it finally looks OK.

Could you please have a quick look at it? Actually, there are more lines of integration tests and documentation then there are real code changes.

Akasurde commented on plugins/modules/vmware_cluster_ha.py:
```diff
@@ -173,6 +173,18 @@
+      - VM storage protection setting for storage failures categorized as All Paths Down (APD).
+      type: str
+      default: 'warning'
+      choices: [ 'disabled', 'warning', 'restartConservative', 'restartAggressive' ]
```
```suggestion
      choices: [ 'disabled', 'warning', 'restartConservative', 'restartAggressive' ]
      version_added: '1.4.0'
```

Akasurde commented on plugins/modules/vmware_cluster_ha.py:
```diff
@@ -173,6 +173,18 @@
+      - VM storage protection setting for storage failures categorized as Permenant Device Loss (PDL).
+      type: str
+      default: 'warning'
+      choices: [ 'disabled', 'warning', 'restartAggressive' ]
```
```suggestion
      choices: [ 'disabled', 'warning', 'restartAggressive' ]
      version_added: '1.4.0'
```

Akasurde reviewed (changes requested):
minor changes, rest lgtm

mariolenz commented:
recheck

mariolenz commented:
recheck

mariolenz commented:
[for the record](https://66d0116e875c83ffba89-aab2e9466e1d4ec3896420b8b8ce1e17.ssl.cf1.rackcdn.com/477/d2c4b12d2b7f647b7fbf70932cefde284dc058f4/check/ansible-test-cloud-integration-vcenter_1esxi_without_nested-python36_1_of_2/7e773a1/)

mariolenz commented:
recheck

mariolenz commented:
recheck

Akasurde commented:
recheck

mariolenz commented:
@Akasurde Thanks for re-triggering the CI. But there seems to be a problem I don't understand, ansible-test-cloud-integration-vcenter_1esxi_without_nested-python36_1_of_2 always crashes. But to me, it doesn't look like it crashes on anything related to my changes to vmware_cluster_ha.

[for the record](https://dashboard.zuul.ansible.com/t/ansible/buildset/58450238f8214b11b42324e025e19671)

mariolenz commented:
recheck

Akasurde commented:
@mariolenz Testcase for Category is failing for some reason and I am not sure. These failures are intermittent so I am speculating that the `vcenter_1esxi_without_nested` environment is lacking enough resources for running these tests. 
Btw, I raised a PR to disable the test for now - https://github.com/ansible-collections/community.vmware/pull/495 (I am not in favor of disabling tests since it has bitten me already).

Let us see what is the failure for this current run.

cc @goneri

mariolenz commented:
recheck

Akasurde reviewed (approved):
LGTM

mariolenz commented:
Thank you @Akasurde for reviewing this PR and for #495 :-)
